stacks/api/function/webhook/src/webhook.service.ts

```diff
@@ -0,0 +1,73 @@
+import {Injectable} from "@nestjs/common";
+import {BaseCollection, DatabaseService} from "@spica-server/database";
+import {Observable} from "rxjs";
+import {Webhook} from "./interface";
+
+@Injectable()
+export class WebhookService extends BaseCollection<Webhook>("webhook") {
+  constructor(database: DatabaseService) {
+    super(database);
+  }
+
+  targets(): Observable<TargetChange> {
+    return new Observable(observer => {
+      const emitHook = (hook: Webhook, kind: ChangeKind) => {
+        const change: TargetChange = {
+          kind: hook.trigger.active ? kind : ChangeKind.Removed,
+          target: hook._id.toHexString()
+        };
+        if (change.kind != ChangeKind.Removed) {
+          change.webhook = {
+            url: hook.url,
+            trigger: hook.trigger
+          };
+        }
+        observer.next(change);
+      };
+      super.find().then(hooks => {
+        for (const hook of hooks) {
+          if (!hook.trigger.active) {
+            continue;
+          }
+          emitHook(hook, ChangeKind.Added);
+        }
+      });
+
+      const stream = this._coll.watch([], {
+        fullDocument: "updateLookup"
+      });
+      stream.on("change", change => {
+        switch (change.operationType) {
+          case "replace":
+          case "update":
+            emitHook(change.fullDocument as Webhook, ChangeKind.Updated);
+            break;
```

Comment:
If we use `findOneAndReplace` in the PUT endpoint, this logic should be inside of replace instead of update case, am I wrong?

Reply:
Nope. We don't care if the operation was an update or replace operation since we do a full clean up after something has changed in the database. So this logic does not and shouldn't rely on the type of the update operation (atomic or non-atomic).

Reply:
Do you have any suggestions about what should be changed here?  and how will it improve the logic here?

Reply:
@htuna07 ?

Reply:
I don't have any suggestions about this logic because it's clear enough. I was thinking about the place of this logic should be `replace` but you explained this situation. That's all :)